### ISSUE TITLE
Why the DalManager needs to open an ObjectContextManager in the Project Tracker example?

### ISSUE THREAD

**Issue by jmarquesh** [MarimerLLC/cslaforum#606]
**Question**
I'm re-reading "Using CSLA 4 -Data Access" and looking at the Project Tracker example. 

I'm wondering why the DalManager needs to open an ObjectContextManager in its constructor:
```
    public DalManager()
    {
      ConnectionManager = ObjectContextManager<PTrackerEntities>.GetManager("PTrackerEntities");
    }
```
If later on each data access code open or uses another:
```
    public List<ProjectDto> Fetch()
    {
      using (var ctx = ObjectContextManager<PTrackerEntities>.GetManager("PTrackerEntities"))
      {...
```

I commented the code that deals with the DbContextManager in the DalManager of one of my projects and seems to be working fine. Even when I save objects with children inside a transaction.

**Version and Platform**
CSLA version: 4.x
OS: Windows
Platform: ASP.NET Core, MVC

**Comment by jonnybee** [comment]
The purpose to avoid bloating your methods with passing the DbContext as a parameter to all children. 
ObjectContextManager puts the aktual DbContext into a LocalThreadStore and uses reference counting sp tha subsequent call to GetManager will return thesame object.

**Comment by jmarquesh** [comment]
Thanks Jonny for your answer. However, I think you're explaining the purpose of the ObjectContextManager which I understand. 

My question is about [this line](https://github.com/MarimerLLC/csla/blob/a6854e474e668381bddca3536e8140775b77e6a9/Samples/ProjectTracker/ProjectTracker.DalEf/DalManager.cs#L27) in **ProjectTracker.DalEf.DalManager**: The DalManager is getting an ObjectContextManager and storing it in a public property named "ConnectionManager". 

Why? 

If later each root object and its children will do the same:

```
using (var ctx = ObjectContextManager<PTrackerEntities>.GetManager("PTrackerEntities"))
      {
```

And the code in ObjectContextManager will take care of getting or disposing the DbContext when needed.

Why is the DalManager also doing it?

I'm sorry if you understood my question the first time and I'm the one confused.

**Comment by jmarquesh** [comment]
Maybe the reason to do it in the DalManager is to keep one DbContext alive for the root object and all the children? If the DalManager doesn't create a DbContext, the one created by the parent will be disposed before the children will be able to use them?

**Comment by rockfordlhotka** [comment]
Yes @jmarquesh, that is exactly what the sample/demo is trying to accomplish.

One goal of the server-side code is to be compositional. Which is to say that a unit of work, or other arbitrary root object, could interact with various other root objects (and their graphs) in ways we can't anticipate today. (you know how code/requirements/software grows and changes over time)

So ensuring that there's this open db connection _regardless_ of the first root object, and _regardless_ of whether several root objects (and their graphs) are activated in various unforeseen orders - that helps ensure that this one connection is reused during the entire life of the server-side invocation.

**Comment by jmarquesh** [comment]
Thanks Rocky. I really appreciate all the time that you guys put into this community.

I'm sorry to insist too much on this, but I'm trying to resolve data access dependencies using an IoC container and I'd love to avoid using the **DalManager**.

There is something that I still don't understand. These paragraphs are from your book "Using CSLA 4 - Data Access":

> Unfortunately it is very common to open multiple connections to a database. If an object graph has a root object that contains a collection of child objects, it is very likely that each object will open a connection to the database to insert, update, or delete the object’s data. Most developers don’t think twice about opening database connections, because .NET implements database connection pooling, so the same connection is reused and performance is good.

>Even with database connection pooling causing the same connection to be reused, the TransactionScope object sees each use of the connection as a separate event and so it thinks the application is opening more than one connection.

As I understand it, the main goal of **DbContextManager** is that a root and children objects do not open several connections. That is accomplished as long as I keep my data access code inside a ```using (var ctx = DbContextManager``` (which I'm doing):
```
//In DalEf
using (var ctx = DbContextManager<GlobalPortalContext>.GetManager())
{
    //Data access code here
}
```

But I still don't get why the **DalManager** has to open the first connection. Why don't we let the first root object to do it? Because maybe it will be disposed before the children or other root objects can use it? It doesn't matter, that is why we have connection pooling. The problem to avoid is opening several connections and, as long as the data access code is inside ```using (var ctx = DbContextManager```, that is accomplished.

What happens if I keep all my data access code in the DalEf inside ```using (var ctx = DbContextManager```, but I don't have a ```using (var dalManager = DalFactory.GetManager())``` in the DataPortal_XYZ of the business objects in the library?

For your reference this is the code of one of my business classes, without using the DalManager:
```
[Transactional(TransactionalTypes.TransactionScope)]
protected override void DataPortal_Insert()
{
	using (var dataScope = ((IBusinessScope)this).Scope.BeginLifetimeScope())
	{
               //The using DalManager used to be here, but now I'm resolving the data access 
               //dependencies using an IoC container

		var dal = dataScope.Resolve<IContactEditDal>();
		using (BypassPropertyChecks)
		{
			var dto = DtoFromThis();
			dal.Insert(dto);
			Id = dto.Id;
		}
		FieldManager.UpdateChildren(this);
	}
}

```
Thanks again for all your help.

**Comment by rockfordlhotka** [comment]
It is fine if the first root object does it - as long as you don't try to later compose that root with other root objects or something. That can happen if you start writing more sophisticated or complex server-side code where the server interacts with various object graphs at one time - so not the more basic CRUD scenario.

That scenario isn't terribly common, but can happen, so when I'm writing demos I try to cover edge cases like that one.

Even then you would typically be using a _unit of work_ pattern, and as long as the UOW implements that top-level `using` block with the data scope you are fine. In my example I was showing one way to implement a solution where it is (virtually) impossible for someone to make a mistake.

**Comment by rockfordlhotka** [comment]
I should add, not knowing exactly what your `dataScope.Resolve` is doing, the _really important thing_ is that you open the db connection ONE TIME ONLY and make sure to keep using that same instance throughout the lifetime of all server-side work.

Your `Resolve` gets an `IContactEditDal` - so it isn't clear who or what is opening the db connection and keeping it open (and reused) by _all the dal objects you might resolve_ as part of a single server-side operation.

In other words, there's something deeper in your code (not shown here) that is opening the db connection. That needs to be a singleton that you inject into each DAL object instance as it is resolved.

**Comment by rockfordlhotka** [comment]
To inject more into this thread (see what I did there 😄 ), you might look at the [Csla.Server.IDataPortalActivator](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/Server/IDataPortalActivator.cs) type, which allows you to use your IoC container to create the actual business object instance.

More importantly, you might also look at [Csla.Server.IInterceptDataPortal](https://github.com/MarimerLLC/csla/blob/master/Source/Csla.Shared/Server/IInterceptDataPortal.cs), which has `Initialize` and `Complete` methods where you can create and tear down your DI scope without having to clutter your root objects with that bit of plumbing.

For example, you can create an autofac scope in `Initialize`, thus making it available until `Complete` is called, when you can tear it down.

@JasonBock wrote a blog post about this some time ago, when we implemented these interfaces to support DI on the server. I think this is the blog post:
https://magenic.com/thinking/abstractions-in-csla

**Comment by rockfordlhotka** [comment]
I do want to specifically answer this statement in your post:

> But I still don't get why the DalManager has to open the first connection. Why don't we let the first root object to do it? Because maybe it will be disposed before the children or other root objects can use it? It doesn't matter, that is why we have connection pooling.

That statement is only true if you don't use any transactions. If you use a transaction, then you need to use the same _instance_ of the database connection for all operations that'll be covered by the transaction.

In other words, database connection pooling can't help you, because you aren't guaranteed the same db conn instance, and even if you got the same instance back, the transaction link would have been broken.

Most apps use transactions when updating multiple tables, and therefore need to ensure the same _connection instance_ is used throughout, so the transaction works as expected.

**Comment by jmarquesh** [comment]
>If you use a transaction, then you need to use the same instance of the database connection for all operations that'll be covered by the transaction.

>In other words, database connection pooling can't help you, because you aren't guaranteed the same db conn instance, and even if you got the same instance back, the transaction link would have been broken.

Oh man, **that bit is what answers my question**. I'm using ```DbContextManager``` in my DalEf code, but I was wondering what was the purpose of having the DalManager to open the first one, instead of letting the first root object do it.

> I should add, not knowing exactly what your dataScope.Resolve is doing...

My ```dataScope.Resolve``` is simply giving me an instance of a class that implements '''IContactEditDal'''. 

Do you remember that trick you do in ```DalManager.GetProvider```, removing the "I" from the interface name and using reflection to create an instance of a type with the name of the interface, minus the "I"? I'm trying to use Autofac to do the same thing. By the way, following the ideas from @JasonBock's blog post that you mentioned: [Abstraction in CSLA](https://magenic.com/thinking/abstractions-in-csla). The problem is that he doesn't have data access code in his post, so I don't have the complete picture.

So, after the resolve, I get an instance of ```ContactEditDal``` in my DalEf project, which have this Insert method:
```
public void Insert(ContactEditDto dto)
{
	using (var ctx = DbContextManager<ContactoContext>.GetManager())
	{
		var newItem = new Contacto.DalEf.Models.Contacts
		{
			ApplicationAccountId = dto.ApplicationAccountId,
			FirstName = dto.FirstName,
			MiddleName = dto.MiddleName,
			LastName = dto.LastName,
			Phone = dto.Phone,
			SendMessages = dto.SendMessages,
			Notes = dto.Notes,
		};
		ctx.DbContext.Contacts.Add(newItem);
		ctx.DbContext.SaveChanges();
		dto.Id = newItem.ContactId;
	}
}
```

As you can see, I'm using a DbContextManager, which should ensure that I don't open several connections. The same pattern is used in the code that persists the children of ```ContactEdit```.


>More importantly, you might also look at Csla.Server.IInterceptDataPortal, which has Initialize and Complete methods where you can create and tear down your DI scope without having to clutter your root objects with that bit of plumbing.

Thanks!

And thank you again for all the time you took to go through my questions.
